api/extension-guides/mcp.md

```diff
@@ -0,0 +1,93 @@
+---
+# DO NOT TOUCH — Managed by doc writer
+ContentId: e655f324-ed0b-452d-aff3-52cdca3978a5
+DateApproved: 05/08/2025
+
+# Summarize the whole topic in less than 300 characters for SEO purpose
+MetaDescription: A guide to registering an MCP server in a VS Code extension.
+---
+
+# MCP servers
+
+Model Context Protocol (MCP) is an open standard that enables AI models to interact with external tools and services through a unified interface. Visual Studio Code can act as an MCP client, which enables users to [access MCP tools in agent mode](/docs/copilot/chat/mcp-servers.md). This article guides you through registering an MCP server in a VS Code extension.
+
+VS Code retrieves MCP server configurations from an `.vscode/mcp.json` file in the workspace, user settings, or can automatically discover them from other tools like Claude Desktop.
+
+VS Code extensions can also register MCP server configurations programmatically to avoid that users need to manually configure them. This is useful if you already have an MCP server and want to register it as part of your extension activation, or if your extension has a dependency on an MCP server.
+
+Instead of using MCP servers to extend the chat functionality, you can also [contribute language model tools](/api/extension-guides/tools.md) directly within your extension. This approach is useful if you want to deeply integrate with VS Code by using extension APIs or to avoid that users have to install and run an MCP server in a separate process.
+
+> [!IMPORTANT]
+> MCP support in VS Code is in preview and the API for registering an MCP server in a VS Code extension is currently in a proposed state.
+
+## Register an MCP server
+
+To register an MCP server in your extension, use the `vscode.lm.registerMcpServerDefinitionProvider` API to provide the [MCP configuration](/docs/copilot/chat/mcp-servers.md#configuration-format) for the server. The API takes a `providerId` string and a `McpServerDefinitionProvider` object.
+
+The `McpServerDefinitionProvider` object has two properties:
+
+- `onDidChangeMcpServerDefinitions`: event that is triggered when the MCP server configurations change.
+- `provideMcpServerDefinitions`: function that returns an array of MCP server configurations (`vscode.McpServerDefinition[]`).
```

Comment:
There is also a `resolve` method now

Reply:
Can you provide more details about this resolve method? Can't figure it out from the example.

Reply:
Does the docblocks on the API help? (If not we should improve them)